src/backend/crew_ai/optimization/feedback_loop.py

```
@@ -0,0 +1,906 @@
+    - Phase 3 adds locator strategies → add _check_locator_rules()
+    - External callers can register custom checkers via register_checker()
+    """
+
+    DEFAULT_CONTRADICTION_RATIO = 0.4
+    DEFAULT_MIN_OBSERVATIONS = 5
+    STALENESS_WINDOW_DAYS = 90  # Anti-patterns not seen in this many days are flagged
+
+    def __init__(
+        self,
+        db_conn,
+        contradiction_threshold: float = None,
```

Ruff (0.15.7)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)